File: 0202-happy-number/0202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;
        
        do {
            slow = cal(slow);
            fast = cal(cal(fast));
        } while (slow != fast);
        
        return slow == 1;
    }
    
    public int cal(int i){
        int ans = 0;
        
        while (i > 0){
            int temp = i % 10;
            ans += temp * temp;
            i /= 10;
        }
        
        return ans;
    }
}

File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        s, f = n, n
        
        def cal(num):
            ans = 0

            while num > 0:
                ans += (num%10)**2
                num //= 10

            return ans
        
        while True:
            s = cal(s)
            f = cal(cal(f))

            if s == f:
                break
        
        return s == 1